pkg/config/identity/api_key.go

```diff
@@ -2,14 +2,61 @@ package identity
 
 import (
 	"context"
+	"fmt"
 
 	"github.com/3scale-labs/authorino/pkg/common"
+	"github.com/3scale-labs/authorino/pkg/common/auth_credentials"
+	ctrl "sigs.k8s.io/controller-runtime"
 )
 
+const (
+	invalidApiKeyMsg  = "the API Key provided is invalid"
+	noApiKeysFoundMsg = "no API Keys were found on the request"
+	authSuccessfulMsg = "Successfully authenticated with the provided API key"
+)
+
+type apiKeyDetails struct {
+	Name                  string            `yaml:"name"`
+	LabelSelectors        map[string]string `yaml:"label_selectors"`
+	authorizedCredentials []string
+}
+
 type APIKey struct {
-	SecretKey string `yaml:"secret_key"`
+	auth_credentials.AuthCredentials
+	apiKeyDetails
+}
+
+var (
+	apiKeyLog = ctrl.Log.WithName("Authorino").WithName("ApiKey")
+)
+
+func NewApiKeyIdentity(name string, labelSelectors map[string]string, authCred auth_credentials.AuthCredentials) (*APIKey, error) {
+	if credentials, err := authCred.GetCredentialsFromCluster(context.TODO(), labelSelectors); err != nil {
+		return nil, err
+	} else {
+		return &APIKey{
+			authCred,
+			apiKeyDetails{
+				name,
+				labelSelectors,
+				credentials,
+			},
+		}, nil
+	}
 }
 
-func (self *APIKey) Call(authContext common.AuthContext, ctx context.Context) (interface{}, error) {
-	return "Authenticated with API key", nil // TODO: implement
+func (apiKey *APIKey) Call(authCtx common.AuthContext, _ context.Context) (interface{}, error) {
+	if reqKey, err := apiKey.GetCredentialsFromReq(authCtx.GetHttp()); err != nil {
+		apiKeyLog.Error(err, noApiKeysFoundMsg)
+		return nil, err
+	} else {
+		for _, secret := range apiKey.authorizedCredentials {
+			if secret == reqKey {
+				return authSuccessfulMsg, nil
```

Comment:
I think here we could return something from the matched secret instead of a fixed string. I understand that this would involve storing more in the `AuthCredentials` struct, but it might be a worthy trade-off. I can see some usage for the secret's annotations, for example, later on in the authorization policies.

Reply:
I left that message knowing we still don't have any usage for any format returned. The idea is whenever we have some specific flow needing more than an error check to change it accordingly.